server-metrics/src/main/scala/org/http4s/server/metrics/Metrics.scala

```diff
@@ -2,102 +2,183 @@ package org.http4s
-                }
-              }
-              .handleErrorWith { cause =>
-                abnormal_terminations.update(elapsed, TimeUnit.NANOSECONDS)
-                Stream.raiseError(cause)
-              }
-          r.map(resp => resp.copy(body = capture(resp.body)))
-        }
-        .leftMap { e =>
-          generalMetrics(method, elapsed)
-          resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
-          service_errors.update(elapsed, TimeUnit.NANOSECONDS)
-          e
-        }
-    }
+    val serviceMetrics = ServiceMetrics(generalServiceMetrics, requestTimers, responseTimers)
 
-    Kleisli { req =>
-      val now = System.nanoTime()
-      active_requests.inc()
-      OptionT {
-        service(req).value.attempt
-          .flatMap(onFinish(req.method, now)(_)
-            .fold[F[Option[Response[F]]]](F.raiseError, F.pure))
+    Kleisli(metricsService[F](serviceMetrics, service)(_))
+  }
+
+  private def metricsService[F[_]: Sync](serviceMetrics: ServiceMetrics, service: HttpService[F])(req: Request[F]): OptionT[F, Response[F]] = {
+    val method = req.method
+    for {
+      now <- OptionT.liftF[F, Long](Sync[F].delay(System.nanoTime()))
+      _ <- OptionT.liftF[F, Unit](Sync[F].delay(serviceMetrics.generalMetrics.active_requests.inc()))
+      resp <- OptionT{
+        service(req)
+          .value
+          .attempt
+          .flatMap(metricsServiceHandler(method, now, serviceMetrics, _))
       }
+    } yield resp
+  }
+
+  private def metricsServiceHandler[F[_]: Sync](method: Method,
+                                                start: Long,
+                                                serviceMetrics: ServiceMetrics,
+                                                e: Either[Throwable, Option[Response[F]]]
+                                               ): F[Option[Response[F]]] = {
+    for {
+      elapsed <- EitherT.liftF[F, Throwable, Long](Sync[F].delay(System.nanoTime() - start))
+      respOpt <- EitherT(e.bitraverse[F, Throwable, Option[Response[F]]](
+        manageServiceErrors(method, elapsed, serviceMetrics).as(_),
+        _.map(manageResponse(method, elapsed, serviceMetrics)).pure[F]
+      ))
+    } yield respOpt
+  }.fold(
+    Sync[F].raiseError[Option[Response[F]]],
+    _.fold(handleUnmatched(serviceMetrics.generalMetrics.active_requests))(handleMatched)
+  ).flatten
+
+  private def manageResponse[F[_]: Sync](
+                                   m: Method,
+                                   elapsed: Long,
+                                   serviceMetrics: ServiceMetrics
+                                 )(response: Response[F]): Response[F] = {
+    val newBody = response.body
+      .onFinalize(
+        incrementCounts(serviceMetrics.generalMetrics.headers_times, elapsed) *>
+        requestMetrics(serviceMetrics.requestTimers, serviceMetrics.generalMetrics.active_requests)(m, elapsed) *>
+        responseMetrics(serviceMetrics.responseTimers, response.status, elapsed)
```

Comment:
These aren't measuring all the way to the end of the body now, right?  That's the difference between `headers_times` and the other timers.  I think these two should be measured to when the finalizer triggers.